libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -20,39 +23,104 @@ import kotlinx.coroutines.delay
 import kotlin.coroutines.resume
 
 /**
  * This class is responsible for refreshing the current direction route's traffic.
  * This does not support alternative routes.
  */
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
 internal class RouteRefreshController(
     private val routeRefreshOptions: RouteRefreshOptions,
     private val routeRefresh: RouteRefresh,
     private val currentIndicesProvider: CurrentIndicesProvider,
     private val routeDiffProvider: DirectionsRouteDiffProvider = DirectionsRouteDiffProvider(),
-    private val localDateProvider: () -> Date
+    private val localDateProvider: () -> Date,
 ) {
 
+    private var state: RouteRefreshStateResult? = null
+        set(value) {
+            if (field == value) return
+            field = value
+            value?.let { nonNullValue ->
+                observers.forEach {
+                    it.onNewState(nonNullValue)
+                }
+            }
+        }
+
+    private val observers = CopyOnWriteArraySet<RouteRefreshStatesObserver>()
+
     internal companion object {
+        @VisibleForTesting
         internal const val LOG_CATEGORY = "RouteRefreshController"
-        private const val FAILED_ATTEMPTS_TO_INVALIDATE_EXPIRING_DATA = 3
+
+        @VisibleForTesting
+        internal const val FAILED_ATTEMPTS_TO_INVALIDATE_EXPIRING_DATA = 3
     }
 
     suspend fun refresh(routes: List<NavigationRoute>): RefreshedRouteInfo {
-        return if (routes.isNotEmpty()) {
-            val routesValidationResults = routes.map { validateRoute(it) }
-            if (routesValidationResults.any { it is RouteValidationResult.Valid }) {
-                tryRefreshingRoutesUntilRouteChanges(routes)
+        try {
+            return if (routes.isNotEmpty()) {
+                val routesValidationResults = routes.map { validateRoute(it) }
+                if (routesValidationResults.any { it is RouteValidationResult.Valid }) {
+                    tryRefreshingRoutesUntilRouteChanges(routes)
+                } else {
+                    val message = joinValidationErrorMessages(routesValidationResults, routes)
+                    onNewState(
+                        RouteRefreshExtra.REFRESH_STATE_FINISHED_FAILED,
+                        "No routes which could be refreshed. $message"
+                    )
+                    waitForever("No routes which could be refreshed. $message")
+                }
             } else {
-                val message = joinValidationErrorMessages(routesValidationResults, routes)
-                waitForever("No routes which could be refreshed. $message")
+                resetState()
+                waitForever("routes are empty")
             }
-        } else {
-            waitForever("routes are empty")
+        } catch (e: CancellationException) {
+            onNewStateIfCurrentIs(
+                RouteRefreshExtra.REFRESH_STATE_CANCELED,
+                current = RouteRefreshExtra.REFRESH_STATE_STARTED,
+            )
```

Comment:
If the current state is not `REFRESH_STATE_STARTED`, shouldn't we reset the state to `null`?

Reply:
I think it must be re-set in can case, let me add the string

Reply:
There're still a lot of corner cases that hard to control, because route refresh controller is not self hosted